### DISCUSSION TITLE
quarto with poetry: The papermill package is required for processing --execute-params

### DISCUSSION THREAD

**Discussion by bryanwhiting** [quarto-dev/quarto-cli#4040] [Q&A] [answered]
I have the `papermill` package installed in my poetry environment, but it's not working
```
poetry run quarto render \
    quarto/_template.qmd \
    -P json_file_path:"xxx" \
    -P data_path:"xxx" \
    --to html
ERROR: 

The papermill package is required for processing --execute-params
```

**Comment by bryanwhiting** [comment, marked as answer]
Answer: i got this working by installing `papermill` in my root directory (by deactivating the conda shell I was in or something). 

E.g.,:
```
deactivate
pip3 install papermill
```
and then execute the command above again and it worked.

Probably just a conflict of environments.

**Reply by johnros** [reply]
Wouldn't this mean that quarto is using the global python installation instead of the active venv?

**Reply by cderv** [reply]
It should not use global python instead of an active venv. We have a doc page on virtual env: https://quarto.org/docs/projects/virtual-environments.html

**Reply by EmiliaJarochowska** [reply]
Is there a way to check if quarto is using a virtual environment or a specific path to python? I have reasons to suspect the latter.
Why?
When I work in an environment where `matplotlib` is clearly installed (e.g. it is listed by `poetry show` and I can import it if I open Python REPL) and try to render a Quarto document from CLI or execute it interactively using either R Studio or VS Code (started from the poetry session using `code .`, I am not sure if there is an equivalent in R Studio), it does not find `matplotlib` as installed.
This is true whether I start the environment using `poetry shell` or execute `poetry run quarto file.qmd`.

**Reply by EmiliaJarochowska** [reply]
A small update on this. When running `quarto render file.qmd` a notebook file is created, `file.ipynb`. Rendering *this one* works fine under conditions as described above. But rendering the original `file.qmd` from which it was created suffers from problems as above.

**Reply by mcanouil** [reply]
Use `quarto check` in the same directory as you run `quarto render/preview`.
And look for the Python "section".

**Reply by EmiliaJarochowska** [reply]
Thank you! This confirmed that quarto is not using python set in the virtual environment. https://quarto.org/docs/projects/virtual-environments.html suggests it would; is there any solution to this? With several python dependencies it is not really possible to work without a virtual environment. Thank you!

**Reply by mcanouil** [reply]
It should use it unless you did not activate your environment and/or not executing Quarto within it.
Far as I am concerned, I have absolutely no issues using Quarto and virtualenv (not Conda) either on Windows, Ubuntu, Debian, or macOS.

**Reply by EmiliaJarochowska** [reply]
Thank you for your patience and help. I've reinstalled the virtual environment, updated all dependencies, and doubled this with `pipenv`. According to [the documentation](https://quarto.org/docs/computations/python.html), Quarto should use Python from PATH:
> Quarto will select a version of Python using the [Python Launcher](https://docs.python.org/3/using/windows.html#python-launcher-for-windows) on Windows or system PATH on MacOS and Linux.

But in my case, `quarto check` from inside of the virtual environment (whether poetry or pipenv):
```
[✓] Checking Python 3 installation....OK
      Version: 3.10.6 (Conda)
      Path: /Users/emilia/.julia/conda/3/bin/python3
      Jupyter: 5.4.0
      Kernels: julia-1.8, julia-1.9, python3
```
I don't know where Quarto gets this because `/Users/emilia/.julia/conda/3/bin/python3` is not in `PATH`. Julia's conda installation is problematic for managing python versions but that's outside of the scope of this discussion. Main thing is that poetry is using Python from `PATH` and Quarto isn't:
```
(automate-output-py3.10) emilia@UU101746 Python % poetry env info

Virtualenv
[...]
Python:     3.10.9
Path:       /Library/Frameworks/Python.framework/Versions/3.10
Executable: /Library/Frameworks/Python.framework/Versions/3.10/bin/python3.10
```
Following [the documentation](https://quarto.org/docs/projects/environment.html) I've set `QUARTO_PYTHON` to be the same as `PATH` in an `_environment` file and thus the same as used by poetry/pipenv. But this didn't change Quatro's behaviour, it still uses the odd path from Julia's conda installation. Any suggestions why setting `QUARTO_PYTHON` doesn't work?
Thank you very much!

**Reply by mcanouil** [reply]
Do you use Python from the Conda environment setup by Julia?
If not, I would suggest to trash it (possibly the whole `.julia` to get a very clean setup).
Then when you setup Julia to work with Jupyter, do so in a virtualenv activated and type "n" when asked to install Python via Conda.

**Reply by EmiliaJarochowska** [reply]
Thank you, removing `.julia` resolved the problem. But the question about Quarto remains: according to the documentation, it should use either Python installation from `PATH` or from the virtual environment. Here it does neither. And setting `QUARTO_PYTHON` in `_environment` doesn't help it - why is Quarto's Python path overriden by some other setting in this case? 

I don't know if my case is unlikely to happen to others but here's how it arose: I don't normally use Python from Julia's Conda environment, it was [the default installed by `PyCall`](https://github.com/JuliaPy/PyCall.jl): 
> By default on Mac and Windows systems, Pkg.add("PyCall") or Pkg.build("PyCall") will use the [Conda.jl](https://github.com/Luthaf/Conda.jl) package to install a minimal Python distribution (via [Miniconda](https://conda.io/projects/conda/en/latest/glossary.html#miniconda-glossary)) that is private to Julia (not in your PATH).

It's possible to skip that installation but that's normally not the default so others might bump across this problem too.

**Reply by mcanouil** [reply]
Unfortunately, I cannot reproduce your situation/setup, so very hard to know what's wrong exactly.

**Reply by cderv** [reply]
Thanks for the report @EmiliaJarochowska 

> I don't know where Quarto gets this because /Users/emilia/.julia/conda/3/bin/python3 is not in PATH

I believe Quarto will default to using Julia bundle python with `Conda.jl` **when julia** is used. This was introduced some time ago at 
* https://github.com/quarto-dev/quarto-cli/pull/1931

So even if not in PATH, Quarto will look in Julia folders for an existing conda environment if set, when Julia is used in the document. 

This is like that since Quarto 1.1 I believe, but we can revisit: Julia configuration with IJulia and Jupyter is not that simple. But IJulia defaults to use Conda.jl so I guess Quarto adapted to that when found. 

So this does not explain why `quarto check` gives you that if you are not in a Julia project / document. 🤔 I wonder if this is a kernels spec order information or maybe PyCall conf is not somehting we support correctly in auto detection - I'll dig into this - Could you open a new reproducible issue with this if possible ? Clean and details step of configuration ? 

> And setting QUARTO_PYTHON in _environment doesn't help it - why is Quarto's Python path overriden by some other setting in this case?

Regarding this, does it **not** work only when using `quarto check` ? I believe this is expected that it does not get picked up by `quarto check` because it currently misses project context awareness (https://github.com/quarto-dev/quarto-cli/issues/6206) 
but it should definitely get picked up by `quarto render`. Did you notice that `quarto render` does not work either ? `_environment.yml` only works in Quarto project context by the way. 

You can also open a new issue about this if you still notice issues

**Reply by EmiliaJarochowska** [reply]
Wow, @cderv, thank you for this extensive answer. It indeed clarifies what has happened:
> I believe Quarto will default to using Julia bundle python with Conda.jl when julia is used.

That is indeed the case here, for some reason this entire repo is detected as a Julia environment, by VS Code as well as by Quarto. I have no idea why that is as I never had any Julia-specific project files not `.jl` files in this repo, but that is of course not a Quarto problem anymore and I [have moved that discussion to Julia Discourse Board](https://discourse.julialang.org/t/julia-environment-falsely-detected-in-a-quarto-python-project/108331).

> Regarding this, does it not work only when using quarto check ? I believe this is expected that it does not get picked up by quarto check because it currently misses project context awareness (https://github.com/quarto-dev/quarto-cli/issues/6206)
but it should definitely get picked up by quarto render. Did you notice that quarto render does not work either ? _environment.yml only works in Quarto project context by the way.

Yes, this also explains (almost) everything, because `quarto render` in a project context when `_environment` is set works even when `quarto check` shows the "wrong" version of Python. The only curious thing that remains is, although this is a quarto project, I rendered individual files for testing. And `quarto render file.qmd` would use the "wrong" Python version from `Conda.jl`, but when it renders, it generates a `file.ipynb` file (even though it doesn't complete because it doesn't have the dependencies set in the virtual environment) and this file is rendered with the **correct** Python installation. 

TL;DR: `quarto render file.qmd` is rendered using a different Python version than `quarto render file.ipynb`, of course in this case `_environment` does not apply.

So really thank you both for helping me with this investigation @cderv and @mcanouil. I don't think this is issue-worthy material, but perhaps a note in the documentation in which order Python versions are chosen by Quarto could be of value, given the number of ways to specify it.

**Reply by cderv** [reply]
> TL;DR: quarto render file.qmd is rendered using a different Python version than quarto render file.ipynb, of course in this case _environment does not apply.

I am not sure I followed everything here but one thing about `quarto render`
- If you do on a .qmd using Jupyter engine, `quarto render` will trigger the notebook execution
- If you do on an existing `.ipynb`, Quarto won't by default execute the cells. 

This is mentioned at  https://quarto.org/docs/projects/code-execution.html#notebooks
> Note that when rendering an .ipynb Quarto will not execute the cells within the notebook by default (the presumption being that you have already executed them while editing the notebook). If you want to execute the cells you can pass the --execute flag to render

This could be a difference if you don't know about this when comparing what happens. 

Sharing just in case it could explain

**Reply by EmiliaJarochowska** [reply]
Thank you and apologies I wasn't clear. I try to explain better. I double-checked and this is still the case.
1. `quarto render file.qmd`: 
- generates `file.ipynb` as a transitional step in rendering
- stops because a wrong Python installation is used and it lacks dependencies
2. `quarto render file.ipynb --execute` works: it uses Python version in the virtual environment, as indicated by including `which python` in the notebook (although I am not sure that executing the shell command necessarily shows which Python was used to execute the notebook).

However, this only occurred when quarto was using `Conda.jl` Python over the one specified in the venv. Now that I removed the `.julia` directory, this problem is null as quarto uses Python selected by poetry.

I'll work on the issue as you suggested but I am still trying to find out why my project is identified as a Julia project, both by Quarto and VS Code. I check all directories in the path to the project for stray Julia files but there weren't any so still no clue.

**Reply by mcanouil** [reply]
What do you mean identified by VSCode as a "Julia project"? There is no such thing in VSCode for Quarto document.

Could it be you did some modification of your VSCode setup?

Did you try Quarto in a Terminal/PowerShell outside of VSCode?

**Reply by EmiliaJarochowska** [reply]
Oh yes, sorry, please see [above](https://github.com/quarto-dev/quarto-cli/discussions/4040#discussioncomment-8017058) for details:
> for some reason this entire repo is detected as a Julia environment, by VS Code as well as by Quarto

I always ran quarto from the terminal, but I just noted that both Quarto and VS Code seem to see it as a Julia project, which probably explains why `Conda.jl` Python was used over `poetry` settings, as explained by @cderv:
> I believe Quarto will default to using Julia bundle python with Conda.jl when julia is used

So principally Quarto behaved correctly, although that would be hard to piece together from documentation. I only continued this dicussion because of the curious fact that different Python seems to have been used to execute `.qmd` and `.ipynb`.

My feeling is that Python path of the virtual environment should have priority over other factors (such as Julia being used), at least from the perspective of replicability, I'll make some experiments with a mixed Quarto Julia & Python project for the issue.

**Reply by cderv** [reply]
> My feeling is that Python path of the virtual environment should have priority over other factors (such as Julia being used), at least from the perspective of replicability, I'll make some experiments with a mixed Quarto Julia & Python project for the issue.

Just as a reminder in this thread is now rather long. Did this happens when the virtual env is activated **before** running quarto render ? 

What happens with `poetry run quarto render ...` ? (https://python-poetry.org/docs/cli/#run)


If you can come with a reproducible example to set up ( maybe using docker, CI space, or even codespaces, DevPod, GitPod, or any other), it would help debug and understand how Quarto does select the python to use and the engine to use. Reminder: Quarto does auto binding, but you can configure the engine to use (https://quarto.org/docs/computations/execution-options.html#engine-binding)

**Comment by atwam** [comment]
Coming here late, having had the same issue.
Your fix, combined with the discussion in https://github.com/quarto-dev/quarto-cli/discussions/5833 helped me solve the problem:

- Having poetry installed, with a proper venv where jupyter is installed
- Running `jupyter kernelspec list` to see my `python3` kernel, and then reading the corresponding json `/home/.../.venv/share/jupyter/kernels/python3/kernel.json`, I got the following content:
```json
{
 "argv": [
  "python",
  "-m",
  "ipykernel_launcher",
  "-f",
  "{connection_file}"
 ],
 "display_name": "Python 3 (ipykernel)",
 "language": "python",
 "metadata": {
  "debugger": true
 }
}
```
 
I think the root of the issue is in the `python` line: when I launch such a kernel, it will usually pick-up the `venv` activated version of `python`. But running `quarto render` will have it get the executable for the kernel (see [exec.ts](https://github.com/quarto-dev/quarto-cli/blob/305bea0c55fac2ca2e3fd218df6f03014ef04e84/src/core/jupyter/exec.ts), which in our case is `python`.
While that executable will be run by quarto (see `jupyter-kernel.ts`), it ends up calling `Deno.run`, which under the hood calls rust's `std::process::Command`. That should take into account the `PATH` variable (according to [its documentation](https://doc.rust-lang.org/std/process/struct.Command.html#method.new)), but it looks like `PATH` was modified somehow, or it find `python` which does not map to the current shell's PATH.

Anyway, changing the kernel.json above to have the full path to the local python shell seems to work. A proper quarto fix would be to query the proper full path of python, by using `which python`, or allowing the jupyter python executable to be set in the options.

**Reply by cderv** [reply]
I am not sure I followed all the specific problem here, but know that quarto 1.5 will correctly check for any activated venv to use this python version. 

So you should try your original way using latest quarto 1.5 prerelease and verify your activated venv in shell is picked up.

**Reply by cscheid** [reply]
> Anyway, changing the kernel.json above to have the full path to the local python shell seems to work. A proper quarto fix would be to query the proper full path of python, by using which python, or allowing the jupyter python executable to be set in the options.

That's not a proper fix. Quarto uses _jupyter_. _Jupyter_ reports that kernel, and we are following what the kernel says. The fix is, unfortunately, for Jupyter+Python package managers to properly manage their kernel.json files.

**Comment by hasibagen** [comment]
I would like to ask how to successfully pass path parameters to quarto's ipynb file. This is my current code. 
````
quarto render test_par.ipynb --to docx -P sample_path="/home/_2024_05_10_15_44_16.csv"
pandoc
to: docx
output-file: test_par.docx
default-image-extension: png
ascii: false

metadata
title: test_par
mainfont: STZhongsong
monofont: STZhongsong
papersize: a4

Output created: test_par.docx

---
title: "test_par"
mainfont: "STZhongsong"
monofont: "STZhongsong"
format: docx
params:
sample_path:
execute:
echo: false
warning: false
eval: true
ascii: false
fig-dpi: 300
papersize: a4
---
#| tags: [parameters]
sample_path = "/home/2024_04_16_16_50_34.csv"
print(sample_path)
````

It doesn't seem to report an error, but there are no parameters passed in either, because the generated path is still the path I wrote in ipynb

[test_par.zip](https://github.com/user-attachments/files/15912085/test_par.zip)

**Reply by mcanouil** [reply]
@hasibagen  Could you properly format your post using code blocks for code and terminal outputs? Thanks.
If your code contains code blocks, you need to enclose it using more backticks, *i.e.*, usually four ` ```` `.

Also, since your question is a bit different, please open a new separate discussion. You can link to this one if appropriate.
Thank you!

**Reply by hasibagen** [reply]
> @hasibagen Could you properly format your post using code blocks for code and terminal outputs? Thanks. If your code contains code blocks, you need to enclose it using more backticks, _i.e._, usually four `` ```` ``.
> 
> Also, since your question is a bit different, please open a new separate discussion. You can link to this one if appropriate. Thank you!

Thank you very much for your suggestion, I have modified the code and raised a new discussion.

https://github.com/quarto-dev/quarto-cli/discussions/10081#discussion-6845141
